[microvm] Adds infrastructure for pause-resume functionality

LucasGdosR opened the pull request:
I add types to every control channel, even though unit types would be enough for some of them. This way, they can be extended cleanly. They all have standardized enumeration names. They're all in the same file, `orchestrator.rs`.

The memory thread and the vCPU threads now have control channels to the main thread.

The `pause_protocol` now includes a step where it sends a message to the memory thread, then it waits for a response from the vCPU thread. Note the memory thread doesn't do anything yet, that's for another PR.

The `resume_microvm` (I guess I should rename it to resume_protocol) now sends a message to the memory thread and waits for a response from the memory thread, then it sends a message to the vCPU thread. Note those messages don't do anything, that's for another PR.

LucasGdosR commented:
These 2 commits should be added on top of the pending commits of https://github.com/nanvix/nanvix/pull/919, that's why I put it as the base branch.

csegarragonz commented:
@LucasGdosR if this PR is pending on another PR, should we mark it as draft?

LucasGdosR commented on src/microvm/src/vmm/microvm/mod.rs:
```diff
@@ -173,13 +189,19 @@ impl Vmm {
-        let memory_thread_tx: Sender<Message> = memory_thread_tx.clone();
-        let memory_thread: JoinHandle<Result<(), anyhow::Error>> =
-            memory_thread::spawn(memory_thread_rx, memory_thread_tx, move || {
+        let memory_thread: JoinHandle<Result<(), anyhow::Error>> = memory::spawn(
```
Fixed.

LucasGdosR commented on src/microvm/src/vmm/hyperlight/mod.rs:
```diff
@@ -277,8 +287,15 @@ impl Vmm {
         // Create a thread that reads from vm_rx and writes to vm_rx2.
-        let memory_thread: JoinHandle<Result<(), anyhow::Error>> =
-            memory_thread::spawn(memory_thread_rx, memory_thread_tx, add_credit);
+        let memory_thread: JoinHandle<Result<(), anyhow::Error>> = memory::spawn(
```
Fixed.

LucasGdosR commented on src/microvm/src/orchestrator.rs:
```diff
@@ -204,22 +279,43 @@ impl Orchestrator {
+                Err(TryRecvError::Empty) => (),
+                Err(TryRecvError::Disconnected) => {
+                    let reason: String = "the vmm has disconnected".to_string();
+                    error!("resume_running_rx.try_recv(): {reason}");
```
Fixed.

LucasGdosR commented on src/microvm/src/orchestrator.rs:
```diff
@@ -271,6 +367,12 @@ impl Orchestrator {
+        // Tell memory thread to write to the kernel
+        self.memory_control_tx.send(MemoryControlCommand::Resume)?;
+        // Wait for confirmation
+        while MemoryControlResponse::ResumeWritten != self.memory_control_rx.recv()? {} // TODO: add timeout check
```
Added the timeout warning instead of leaving that `TODO`.

LucasGdosR commented:
I have dealt with all of copilot's valid comments. This should be good to go. EDIT: typo in hyperlight. Now should be good to go.

LucasGdosR commented on src/user/snapshot-rust-nostd/src/main.rs:
```diff
@@ -0,0 +1,56 @@
+    let mut buffer = [0u8; 32];
+
+    loop {
+        if count % MULTIPLIER == 0 {
+            let count_str = count_to_string(count / MULTIPLIER, &mut buffer);
+            let _ = debug::debug(count_str.as_ptr(), count_str.len());
+        } else if count == 1 << 63 {
+            break;
```
Either way works. Even if this was an infinite loop (which it isn't), it would serve its purposes of pausing-resuming, and later of saving snapshots and loading them.

ppenna replied:
@LucasGdosR, the break condition for the loop is not clear. Could you please rewrite it? People often times rely on tests/examples to learn about the code base.

I still need to review this PR, but I also found it hard to understand what `MULTIPLIER` means.

LucasGdosR replied:
It's now a long for-loop that prints every ITERATION_STEPS iterations.

ppenna commented on src/microvm/src/memory_thread.rs:
```diff
@@ -79,6 +93,28 @@ where
+                Err(TryRecvError::Disconnected) => {
+                    // When the guest finishes , the vCPU thread will disconnect from this
+                    // thread. This situation is normal and should not create an error log.
+                    debug!("memory_thread(): channel has been disconnected");
```
This should be tagged as `spawn()`.

ppenna commented on src/microvm/src/memory_thread.rs:
```diff
@@ -79,6 +93,28 @@ where
+                Ok(MemoryControlCommand::Pause) => {
+                    trace!("pause()");
+                    crate::timer!("vm_pause");
+                    pause_microvm()?;
```
If we fail, we should send a message through the control channel, so it can handle graceful shutdown.

LucasGdosR replied:
Note, however, that the VMM thread doesn't wait for a response from the memory thread at this point. Instead, it waits for a response from the vCPU thread. I will add to the timeout loop in `pause_protocol` a check for control responses from the memory thread.

ppenna commented on src/microvm/src/memory_thread.rs:
```diff
@@ -79,6 +93,28 @@ where
+                Ok(MemoryControlCommand::Resume) => {
+                    trace!("resume()");
+                    crate::timer!("vm_resume");
+                    resume_microvm()?;
```
If we fail, we should send a message through the control channel, so it can handle graceful shutdown.

ppenna commented on src/microvm/src/orchestrator.rs:
```diff
@@ -204,22 +282,43 @@ impl Orchestrator {
+                Err(TryRecvError::Empty) => (),
+                Err(TryRecvError::Disconnected) => {
+                    let reason: String = "the vmm has disconnected".to_string();
+                    error!("vcpu_control_rx.try_recv(): {reason}");
```
This tag should be "pause_protocol()"

ppenna commented on src/user/snapshot-rust-nostd/src/main.rs:
```diff
@@ -0,0 +1,56 @@
+
+#![no_std]
+#![no_main]
+
```
Can we have a module-level documentation that explains how this test work?

ppenna reviewed (changes requested):
I left some minor comments here and there.

LucasGdosR commented:
Just so I don't leave you hanging for long, I'll get back to this on Thursday.

LucasGdosR commented:
Something weird happened with the benchmark test, but the others have passed. Did I do something wrong?

ppenna commented:
> Something weird happened with the benchmark test, but the others have passed. Did I do something wrong?

@LucasGdosR no need to worry for now I believe it is a glitch related to @csegarragonz's PR that got merged.

ppenna commented on src/microvm/src/memory_thread.rs:
```diff
@@ -79,6 +93,32 @@ where
                     // No message available.
                 },
             }
+            match control_rx.try_recv() {
+                Ok(MemoryControlCommand::Pause) => {
+                    trace!("pause()");
+                    crate::timer!("vm_pause");
+                    if pause_microvm().is_err() {
+                        control_tx.send(MemoryControlResponse::PauseError)?;
+                    }
+                },
+                Ok(MemoryControlCommand::Resume) => {
+                    trace!("resume()");
+                    crate::timer!("vm_resume");
+                    if resume_microvm().is_err() {
+                        control_tx.send(MemoryControlResponse::ResumeError)?;
+                    }
+                    control_tx.send(MemoryControlResponse::ResumeWritten)?;
+                },
+                Err(TryRecvError::Disconnected) => {
+                    // When the guest finishes , the vCPU thread will disconnect from this
+                    // thread. This situation is normal and should not create an error log.
+                    debug!("spawn(): channel has been disconnected");
+                    break Ok(());
+                },
+                Err(TryRecvError::Empty) => {
+                    // No message available.
+                },
+            }
```
Trace tags should contain the name of the function. For this case, all trace statements should use "spawn()". Rationale: this is used to pin-point the module and function where the log comes from. Since in the same module function names are unique, there is no way of having any ambiguity.

ppenna commented on src/microvm/src/orchestrator.rs:
```diff
@@ -204,22 +284,55 @@ impl Orchestrator {
+            // Log a warning and increment the counter every TIMEOUT_WARNING_INTERVAL_IN_MS ms.
+            let elapsed_time: usize = start.elapsed().as_millis() as usize;
+            if elapsed_time > TIMEOUT_WARNING_INTERVAL_IN_MS * counter {
+                warn!(
+                    "{}ms have passed waiting for `ResumeMicroVm` message",
+                    TIMEOUT_WARNING_INTERVAL_IN_MS * counter
+                );
+                match self.memory_control_rx.try_recv() {
```
Please tag this warn message with "pause_protocol()" as well.

ppenna commented on src/microvm/src/orchestrator.rs:
```diff
@@ -204,22 +284,55 @@ impl Orchestrator {
+                );
+                match self.memory_control_rx.try_recv() {
+                    Ok(MemoryControlResponse::PauseError) => todo!(), // TODO: graceful shutdown
+                    Ok(_) => unreachable!(
```
Please match this case, even if it is an empty value.

LucasGdosR replied:
I assume the issue is not with `unreachable!()`, but with `Ok(_)`. I explicitly match all enum variations, but keep some of them as unreachable.

ppenna commented on src/microvm/src/orchestrator.rs:
```diff
@@ -204,22 +284,55 @@ impl Orchestrator {
+            }
+            // Log a warning and increment the counter every TIMEOUT_WARNING_INTERVAL_IN_MS ms.
+            let elapsed_time: usize = start.elapsed().as_millis() as usize;
+            if elapsed_time > TIMEOUT_WARNING_INTERVAL_IN_MS * counter {
+                warn!(
+                    "{}ms have passed waiting for `ResumeMicroVm` message",
+                    TIMEOUT_WARNING_INTERVAL_IN_MS * counter
+                );
+                match self.memory_control_rx.try_recv() {
+                    Ok(MemoryControlResponse::PauseError) => todo!(), // TODO: graceful shutdown
+                    Ok(_) => unreachable!(
+                        "PauseError is the only message that can be sent at this point."
+                    ),
+                    Err(TryRecvError::Empty) => (),
+                    Err(TryRecvError::Disconnected) => {
+                        let reason: String = "the vmm has disconnected".to_string();
+                        error!("pause_protocol(): {reason}");
+                        anyhow::bail!(reason)
+                    },
+                }
+                counter += 1;
+            }
+        }
```
We should have a back-off exponential timeout to abort the protocol. I am fine addressing this in a different PR. Could you please create an issue and link it here?

LucasGdosR replied:
https://github.com/nanvix/nanvix/issues/943

ppenna reviewed (changes requested):
Please decouple these PR in two: this one, and a first one where you just rename variables.

LucasGdosR commented:
> Please decouple these PR in two: this one, and a first one where you just rename variables.

I went ahead and decoupled into more PRs than that. https://github.com/nanvix/nanvix/pull/934 is for renaming, https://github.com/nanvix/nanvix/pull/935 is for the user program, and I'll add a third one for the rest of the infrastructure once the renaming has been merged into dev.

LucasGdosR committed:
[microvm] E: Adding codepaths to issue pause-resume commands

Added types to every control channel, even though unit types would be enough for some of them. This way, they can be extended cleanly. They all have standardized enumeration names. They're all in the same file, `orchestrator.rs`.

The memory thread and the vCPU threads now have control channels to the main thread.

The `pause_protocol` now includes a step where it sends a message to the memory thread, then it waits for a response from the vCPU thread. Note the memory thread doesn't do anything yet, that's for another PR.

The `resume_microvm` (I guess I should rename it to resume_protocol) now sends a message to the memory thread and waits for a response from the memory thread, then it sends a message to the vCPU thread. Note those messages don't do anything, that's for another PR.

ppenna commented on src/microvm/src/memory_thread.rs:
```diff
@@ -64,15 +79,41 @@ where
+            }
+            match control_rx.try_recv() {
+                Ok(MemoryControlCommand::Pause) => {
+                    trace!("pause()");
```
You missed this trace statement, or perhaps it is dangling.

ppenna commented on src/microvm/src/memory_thread.rs:
```diff
@@ -64,15 +79,41 @@ where
+                    }
+                },
+                Ok(MemoryControlCommand::Resume) => {
+                    trace!("resume()");
```
Same here

ppenna commented on src/microvm/src/memory_thread.rs:
```diff
@@ -64,15 +79,41 @@ where
+                    trace!("pause()");
+                    crate::timer!("vm_pause");
+                    if pause_microvm().is_err() {
+                        control_tx.send(MemoryControlResponse::PauseError)?;
```
Please log an error before failing.

ppenna commented on src/microvm/src/orchestrator.rs:
```diff
@@ -204,9 +284,47 @@ impl Orchestrator {
-        // This TODO requires pausing the vCPU and a control plane communication with linuxd
+        // TODO: tell linuxd to flush (Running -> Flushing)
+        // This TODO requires control plane communication with linuxd
+        self.memory_control_tx.send(MemoryControlCommand::Pause)?;
```
Please log an error before failing

LucasGdosR replied:
I have added this log. However, I don't get it. Whenever the `pause_protocol()` function gets called (the one wrapping this `send()` call), there's error handling, so there's no way to throw an uncaught error. I also see quite a few instances of `send()?` throughout the codebase. Why does this particular `send()` need logging? Or should all other `send()?` calls also have error logging?

ppenna reviewed (changes requested):
The number of TODOs in the main code is growing rapidly. For each group of them create and link an issue on the TODO comment.

LucasGdosR committed:
[microvm] Linking TODO with issue

LucasGdosR committed:
[microvm] E: Adding error logging and removing tracing

csegarragonz commented on src/microvm/src/vmm/hyperlight/mod.rs:
```diff
@@ -421,3 +441,13 @@ fn consume_credit() -> Result<()> {
         .ok_or(anyhow::anyhow!("VMEM is not initialized"))?
 }
+
+// Requests the kernel to pause the virtual machine's execution by writing to a specific register.
```
Function block comments should follow the syntax:

```rust
///
/// # Description
///
fn ...
```

LucasGdosR replied:
Fixed. I was under the impression that `hyperlight/mod.rs` was more relaxed in this regard, but I guess it's just because it's an older portion of the codebase. Broken windows theory, you know?

LucasGdosR commented on src/microvm/src/orchestrator.rs:
```diff
@@ -204,18 +282,60 @@ impl Orchestrator {
+            let elapsed_time: usize = start.elapsed().as_millis() as usize;
+            if elapsed_time > TIMEOUT_WARNING_INTERVAL_IN_MS * counter {
+                warn!(
+                    "pause_protocol(): {}ms have passed waiting for `ResumeMicroVm` message",
```
Fixed.

LucasGdosR commented on src/microvm/src/memory_thread.rs:
```diff
@@ -64,15 +79,58 @@ where
+                        if let Err(e) = control_tx.send(MemoryControlResponse::PauseError) {
+                            let reason: String =
+                                format!("error sending `PauseError` to the VMM (error={e:?})");
+                            error!("spawn: {reason}");
```
Fixed.

LucasGdosR commented on src/microvm/src/memory_thread.rs:
```diff
@@ -64,15 +79,58 @@ where
+                        if let Err(e) = control_tx.send(MemoryControlResponse::ResumeError) {
+                            let reason: String =
+                                format!("error sending `ResumeError` to the VMM (error={e:?})");
+                            error!("spawn: {reason}");
```
Fixed.

csegarragonz commented on src/microvm/src/orchestrator.rs:
```diff
@@ -268,6 +388,36 @@ impl Orchestrator {
+        // Wait for confirmation
+        let start: Instant = Instant::now();
+        let mut counter: usize = 1;
+        while match self.memory_control_rx.try_recv() {
```
I find this while loop very hard to parse, particularly because the boolean condition evaluation takes many lines itself. Could we instead do something like:

```rust
loop {
    match self.memory_control_rx.try_recv() {
        Ok(MemoryControlResponse::ResumeWritten) => break,
        ...
        Err(TryRecvError::Empty) => continue,
    }

    // Sleep for a bit
}
```

LucasGdosR replied:
I simplified it using `loop` and `break`, as per your and copilot's suggestions. However, I think spin-polling is more appropriate than sleeping in this case. All the memory thread does when receiving that message is writing a u32 to memory and send back that message. It shouldn't spin for long, and it's likely better than going through the scheduler after sleeping.

csegarragonz commented on src/microvm/src/orchestrator.rs:
```diff
@@ -61,6 +61,78 @@ enum State {
+pub enum MemoryControlResponse {
+    PauseError,
+    ResumeError,
+    ResumeWritten,
```
I wonder if we could use something like `Success` that captures both when `Pause` and `Resume` succeed.

LucasGdosR replied:
There's no `PauseSuccess`, as the VMM does not wait for a reply from the memory thread, only from the vCPU thread. We could use a generic `Error` in place of both `PauseError` and `ResumeError` at some point.

csegarragonz reviewed (commented):
Watch out, commit `cceb62225aadd9b4a3070af11aec1d2102177b50` does not follow the commit guidelines.

LucasGdosR committed:
[microvm] E: Implementing suggestions

LucasGdosR commented on src/microvm/src/orchestrator.rs:
```diff
@@ -268,6 +388,36 @@ impl Orchestrator {
+                    "{}ms have passed waiting for `ResumeWritten`",
+                    TIMEOUT_WARNING_INTERVAL_IN_MS * counter
+                );
+                counter += 1;
```
Carlos also pointed this out, so I changed it accordingly.

LucasGdosR commented on src/microvm/src/orchestrator.rs:
```diff
@@ -204,18 +282,60 @@ impl Orchestrator {
+                        "PauseError is the only message that can be sent at this point."
+                    ),
+                    Ok(MemoryControlResponse::ResumeWritten) => unreachable!(
+                        "PauseError is the only message that can be sent at this point."
```
Added the extra context to the `unreachable!()` strings.

LucasGdosR commented on src/microvm/src/memory_thread.rs:
```diff
@@ -64,15 +79,58 @@ where
+                Err(TryRecvError::Disconnected) => {
+                    // When the guest finishes , the vCPU thread will disconnect from this
+                    // thread. This situation is normal and should not create an error log.
+                    debug!("spawn(): channel has been disconnected");
```
Fixed.

ppenna commented:
@LucasGdosR would you mark conversations as resolved once you address them?

Also, let's improve CI throughput by following this strategy:

1. Open Draft PR
2. Ask for reviews
3. If changes are not approved, work on the reviews, push changes and go to step 2.
4. If changes are approved, change draft PR to regular PR and let CI run.
5. If CI passes, ask for reviews.
6. If changes are not approved, change PR to draft, work on changes, push changes and go to step 2.
7. If changes are approved, wait for PR to be merged.

Also, before pushing changes, consider amending commits to keep your commit history clean.

LucasGdosR commented:
> @LucasGdosR would you mark conversations as resolved once you address them?
> 
> Also, let's improve CI throughput by following this strategy:
> 
> 1. Open Draft PR
> 2. Ask for reviews
> 3. If changes are not approved, work on the reviews, push changes and go to step 2.
> 4. If changes are approved, change draft PR to regular PR and let CI run.
> 5. If CI passes, ask for reviews.
> 6. If changes are not approved, change PR to draft, work on changes, push changes and go to step 2.
> 7. If changes are approved, wait for PR to be merged.
> 
> Also, before pushing changes, consider amending commits to keep your commit history clean.

Roger that. I will change back and forth between draft and regular PRs and avoid letting CI run without first getting reviews.

LucasGdosR commented:
> @LucasGdosR would you mark conversations as resolved once you address them?
> 
> Also, let's improve CI throughput by following this strategy:
> 
> 1. Open Draft PR
> 2. Ask for reviews
> 3. If changes are not approved, work on the reviews, push changes and go to step 2.
> 4. If changes are approved, change draft PR to regular PR and let CI run.
> 5. If CI passes, ask for reviews.
> 6. If changes are not approved, change PR to draft, work on changes, push changes and go to step 2.
> 7. If changes are approved, wait for PR to be merged.
> 
> Also, before pushing changes, consider amending commits to keep your commit history clean.

@ppenna to be clear, I have addressed the points brought up in review, and CI has passed. From my understanding this PR is waiting to be merged. In case there's something else in the way, I'll be happy to address it :)

ppenna reviewed (approved):
Thank you for working on these changes, @LucasGdosR
